runtime/src/main/kotlin/kotlin/native/concurrent/Lazy.kt

```diff
@@ -6,55 +6,49 @@
-import kotlin.native.internal.NoReorderFields
 
-@SymbolName("Konan_ensureAcyclicAndSet")
-private external fun ensureAcyclicAndSet(where: Any, index: Int, what: Any?): Boolean
-
-@SymbolName("ReadHeapRefNoLock")
-internal external fun readHeapRefNoLock(where: Any, index: Int): Any?
-
-@NoReorderFields
 internal class FreezeAwareLazyImpl<out T>(initializer: () -> T) : Lazy<T> {
-    // IMPORTANT: due to simplified ensureAcyclicAndSet() semantics fields here must be ordered like this,
-    // as an ordinal is used to refer a field.
-    private var value_: Any? = UNINITIALIZED
-    private var initializer_: (() -> T)? = initializer
+    private val value_ = FreezableAtomicReference<Any?>(UNINITIALIZED)
+    private val initializer_ = FreezableAtomicReference<(() -> T)?>(initializer)
     private val lock_ = Lock()
 
     override val value: T
         get() {
             if (isFrozen) {
                 locked(lock_) {
-                    // Lock is already taken above.
-                    var result = readHeapRefNoLock(this, 0)
+                    var result = value_.value
                     if (result !== UNINITIALIZED) {
-                        assert(result !== INITIALIZING)
+                        if (result === INITIALIZING) {
+                            throw IllegalStateException("Recursive lazy computation")
+                        }
                         @Suppress("UNCHECKED_CAST")
                         return result as T
                     }
                     // Set value_ to INITIALIZING.
-                    ensureAcyclicAndSet(this, 0, INITIALIZING)
-                    result = initializer_!!().freeze()
-                    // Set value_.
-                    if (!ensureAcyclicAndSet(this, 0, result)) {
-                        throw InvalidMutabilityException("Setting cyclic data via lazy in $this: $result")
-                    }
-                    // Do not clear initializer_ reference, as it may break freezing invariants and zero out
-                    // still valid object. It seems to be safe only in case when `this` is not reachable from
-                    // initializer.
+                    value_.value = INITIALIZING
+                    result = initializer_.value!!().freeze()
+                    // Set value_ to actual one.
+                    value_.value = result
+                    // Clear initializer.
+                    initializer_.value = null
                     @Suppress("UNCHECKED_CAST")
                     return result as T
                 }
             } else {
-                var result: Any? = value_
-                if (result === UNINITIALIZED) {
-                    result = initializer_!!()
-                    if (isFrozen)
-                        throw InvalidMutabilityException("$this got frozen during lazy evaluation" )
-                    value_ = result
-                    initializer_ = null
+                var result: Any? = value_.value
```

Comment:
The entire code block has little difference from the one under the lock above.
Isn't it possible to extract it to a function instead?